### ISSUE TITLE
uBlock as an HTTP proxy

### ISSUE THREAD

**Issue by Zearin** [uBlock-LLC/uBlock#1475]
Several years ago, I used to use [GlimmerBlocker](http://glimmerblocker.org).  As a system-wide HTTP proxy, it touted benefits over traditional ad blocking extensions: 
- it was system wide (rather than browser-specific)
- it offered incredible fine-tuning of requests (including modification, which I recognize uBlock deliberately eschews for performance)
- it did not rely on SIMBL “extensions” or “hacks”, which was the standard method for Safari users to block ads at the time (before Safari extensions existed).

At the same time, I lamented some of its drawbacks:
- it was written in Java (which is seldom installed on Macs these days)
- its development slowed significantly
- input for future development was not easy (only 1 developer, only reachable through email, no community discussion)
- for technical reasons, it was unable to block (or do anything with) HTTPS requests

When GlimmerBlocker was first released, most ads were sent over HTTP.  GlimmerBlocker’s website even banks on this, stating that ads were not likely to use HTTPS because of the additional overhead.  Back then, it was true.  Today…not so much. :-/

---

Today, I use uBlock and absolutely love it, and don’t really keep up with GlimmerBlocker any more.  But in spite of its large HTTPS shortcoming, there was something I miss _dearly_ about having system-wide ad blocking.  

**Would it be (_theoretically_) possible to adapt uBlock into some kind of system-wide HTTP(S?) proxy?**  

Obviously, the scope would probably require a separate project.  But think of it:

Rather than protecting individual browsers on each computer, I love the idea of uBlock protecting the **whole** computer.  (Or better yet, having a dedicated machine running uBlock to protect an entire **network**.)

---

If this is a naïve request, sorry about that; I’ve been immersed in the Web since the late 90s, but I’ve never been great with networking itself.  I just do my best.

Still, I **love** uBlock’s user-empowering mission, and its dedication to integrity and performance.  I just want it to protect as much as possible.  

Thanks for reading this long post!, and may The Source™ be with you.

**Comment by floam** [comment]
Glimmer blocker was a lot like privoxy but with a nice Mac UI.

You're what to do is talk a very close look at [Privoxy](http://www.privoxy.org). It's a multi-platform, ad-blocking, privacy protecting web proxy. It has been around forever and lots of people sue it, and are doing different kinds of things extending upon it or forking it and scratching some itch with it. Privoxy was a part of the Tor Browser Bundle the last time I installed it. And  it is not an abandoned project (there is [some activity](http://ijbswa.cvs.sourceforge.net/viewvc/ijbswa/current/gateway.c?view=log&sortby=date&pathrev=MAIN) on their, ahem, SourceForge... [CVS repository](http://ijbswa.cvs.sourceforge.net/viewvc/ijbswa/current/?sortby=date&pathrev=MAIN).

**Comment by floam** [comment]
Seriously though. Want to block ads?
- Just use uBlock everywhere.
- Keep using a HTTP proxy if you like.

**Comment by Zearin** [comment]
Daaaamn, @floam…thank you for _the_ reply of unsurpassed detail!

I read everything, but I’ll limit my response to the most important points.

> Do check out Privoxy. It's a multi-platform, ad-blocking, privacy protecting HTTP proxy. It has been around forever and people use it, lots of projects adapt it for various things. 

Cool!  I’ll check it out.

> Still, use this in conjunction with uBlock..

Oh, of course!  

You mentioned using other stuff in conjunction with uBlock a couple of times.  I couldn’t agree more.  

I hope I didn’t give the impression that I wanted to replace uBlock as a browser extension with my suggestion. Rather, I just longed for the system-wide protection offered by GlimmerBlocker, and wondered if it would be possible to abstract out a piece of uBlock that could take care of the fastest, easiest blocking on a system-wide level.

> It has you trust their certificate to make it work. Probably a nice guy but he can man-in-the-middle you into thinking he is your bank website if he decided to or he were hacked.

Ew.  No.  I do not trust this!

> Additionally install a HOSTS file. Update it semi-often, uBlock has a good reference list. There are tools to update them for you.

Oh…see, this is why I said “I’ve never been great with networking” in my original post.  I think I remember hearing about a HOSTS file, but I don’t know what it is.  I’ll start reading up after I finish this post. :)

> If you want to do this to the "whole network" you could use dnsmasq on your router and DNS blackhole bad domains with that same list of bad hostnames. 

More network ignorance here!  :)  I don’t really understand this.  I can guess that it sounds like adding a blacklist to my router, but I don’t really know how to run custom software on a router; all I know how to do is use the shameful, frustrating, horrible web UI to poke around the router settings (many of which I also don’t understand).

> This project makes it pretty easy to do this if you've got a Linux machine or perhaps your router/gateway you can use as a DNS server. There are lots of ways to do this.

I know gateways have something to do with “splitting” a single IP address into many on a subnetwork, but not much more than that.  (And I wouldn’t be surprised if that description had technical inaccuracies as well.)  But 

> Don't rely on HOSTS and a proxy alone for basic privacy/adblocking. In certain circumstances things could become slower and lots of stuff will be missed. If an ad has an IP address in mind for a resource, some app does a DNS lookup itself on an outside server, it's no problem to get around your settings.

I believe I understand this.  

Use HOSTS is for basic, straightforward stuff, but it’s not foolproof.  Got it.

After I learn about HOSTS files, I hope to use them soon.  Maybe I could even write a script to `curl` the adblock feeds and keep it up-to-date automatically…(?).  Is that feasible?

> Since you use a Mac you may enjoy Little Snitch to catch apps phoning home and possibly unusual activity.

I **do** use Little Snitch!  It’s pretty fantastic.  

Its UI isn’t set up to make it easy to block ever-changing lists of ad servers, though.  I’ve blocked a few ads with Little Snitch, but I don’t know of a way to have it do something like subscribe to a feed of blocking rules, for example.

**Comment by anatasiajp** [comment]
> > It has you trust their certificate to make it work. Probably a nice guy but he can man-in-the-middle you into thinking he is your bank website if he decided to or he were hacked.

What ? With your personal self-signed CA.crt, only you can decrypt your HTTPS content, so no other can read it, that Man in the middle method is 100% safe, you are wrong. I'm writting a local proxy like Privoxy but support HTTPS using that technology so I'm sure that is safe.

> > Don't rely on HOSTS and a proxy alone for basic privacy/adblocking. In certain circumstances things could become slower and lots of stuff will be missed.

That only true with HOSTS, but with local proxy you even can have a better ad filtering performance, much faster than any kind of browser's AdBlock software by abusing multiprocessing.

Okay, to the author of this topic:
- If you want a local filtering software that can filter HTTPS, try Proxydomo (the best at the moment), or you can try  GlimmerBlocker, Privoxy, Proximitron with ProxHTTPSProxy, this will help those three filter HTTPS website.

**Comment by floam** [comment]
> What ? With your personal self-signed CA.crt, only you can decrypt your HTTPS content, so no other can read it, that Man in the middle method is 100% safe, you are wrong. I'm writting a local proxy like Privoxy but support HTTPS using that technology so I'm sure that is safe.

A personal self-signed certificate is pretty much fine. Did I misunderstand something?

Their [instructions](https://github.com/epitron/mitm-adblock) as I understand them send you over to https://mitm.it and accept THEIR certificate. I'm confused... I didn't go further with this but are you saying that page actually helps you generate your own cert?

**Comment by floam** [comment]
OK, looked deeper. I had not realized that mitm.it was a part of the mitmproxy project. They're kosher and I had not realized that after running the proxy, by navigating there you'll not get what I saw but a tool to help users get certs going under their own CA.

Still don't quite understand why their own cert on that real-hosted version of that page needs to hit you with a domain mismatch for this to work the way I learned it does now. I guessed their first step was to have you always trust that cert, then next page helps you install their CA.

I was wrong, and I am sorry for not looking into that deeper and regret if my advice influenced any decision making.
